streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -275,11 +298,16 @@ private void restoreState(final StateRestoreCallback stateRestoreCallback,
             stateRestoreListener.onRestoreStart(topicPartition, storeName, offset, highWatermark);
             long restoreCount = 0L;
 
+            final RecordDeserializer recordDeserializer = deserializers.get(topicPartition.topic());
+
             while (offset < highWatermark) {
                 try {
                     final ConsumerRecords<byte[], byte[]> records = globalConsumer.poll(pollTime);
                     final List<ConsumerRecord<byte[], byte[]>> restoreRecords = new ArrayList<>();
                     for (final ConsumerRecord<byte[], byte[]> record : records.records(topicPartition)) {
+                        if (recordDeserializer != null && recordDeserializer.deserialize(processorContext, record) == null) {
```

Comment:
> It seems like global stores can only be populated from input topics, is that right? In that case, they probably never need a changelog and just instead always populate and restore from the input. If that's true, then why do global stores need a concept of "restore" at all? It seems at first brush like populate and restore are functionally equivalent.

This problem is not specific to global stores, I just did this part of the implementation first and would extend it to the local stores afterwards.
The point is that any unserializable record in the source topic can be handled during normal operations but will be loaded in the store upon restore, potentially causing a problem when accessing the record from the store.
And without this kind of protection there is not a lot you can do except try to clean up the topic itself or try to catch deserialization problems anywhere the store is accessed.

> Maybe the restore path is more efficient for some reason, like it polls in larger batches or something?

It would be a lot more efficient if not every record had to be inspected, although I didn't change the batch loading (hopefully). I only try to deserialize the record and skip it when it fails.

Reply:
When using the LogAndContinueExceptionHandler the deserialize returns null for the whole ConsumerRecord.
The tombstone should have a key set and value null, right?
I will try to add a test for the tombstone case, just to be sure.